metaflow/plugins/aws/awslambda/lambda_runner.py

```diff
@@ -0,0 +1,291 @@
+
+class LogTailer(Thread):
+
+    _stop_flag = False
+
+    def __init__(
+        self,
+        stdout_location,  # type: str
+        stderr_location,  # type: str
+        echo,  # type: Callable[[str, bool], None]
+    ):
+        """
+        Parameters
+        ----------
+        echo : Callable[str]
+            Printer function to call for each log line
+        """
+        self._stdout_tail = S3Tail(stdout_location)
+        self._stderr_tail = S3Tail(stderr_location)
+
+        self._echo = echo
+        super(LogTailer, self).__init__()
+        self.daemon = False
+
+    def stop_tail(self):
+        self._stop_flag = True
+
+    def _print_available_from_s3(self):
+        try:
+            for line in self._stdout_tail:
+                line = refine(line, prefix="")
+                self._echo(line.strip().decode("utf-8", errors="replace"), False)
+            for line in self._stderr_tail:
+                line = refine(line, prefix="")
+                self._echo(line.strip().decode("utf-8", errors="replace"), True)
+        except Exception as ex:
+            self._echo("[ temporary error in fetching logs: %s ]" % ex, True)
+
+    def run(self):
+        delayed_update_while(
+            condition=lambda: not self._stop_flag,
+            update_fn=self._print_available_from_s3,
+        )
+        self._print_available_from_s3()
+
+
+def lambda_name(memory_mb, timeout_seconds, image_uri):
+    """ Construct lambda name based on hashed resources """
+    res_hash = hashlib.sha256(
+        str(memory_mb).encode("utf8")
+        + str(timeout_seconds).encode("utf8")
+        + image_uri.encode("utf8")
+    ).hexdigest()[:10]
+
+    # We append unhashed resource values here for a bit more debuggability
+    return "-".join(["mflambda", res_hash, str(memory_mb), str(timeout_seconds)])
+
+
+def ensure_lambda(
+    client,
+    flow_name,
+    step_name,
+    timeout_seconds,  # type: int
+    memory_mb,  # type: Optional[int]
+    role_arn,
+    image_uri,
+    force_fetch_image,  # type: bool
+    echo,  # type: Callable[[str, bool], None]
+):
+    # type: (...) -> str
+    """
+    Make sure Lambda for the step exists. If it does, but resources or timeout
+    configuration is not what we expect, update the lambda configuration.
+
+    Return lambda ARN.
+    """
+
+    if memory_mb is not None:
+        memory_size = memory_mb
+    else:
+        memory_size = 2048
+
+    # Compute lambda name
+    name = lambda_name(memory_mb, timeout_seconds, image_uri)
+
+    # Since multiple tasks in a foreach branch reuse the same Lambda, they all might
+    # get to this point simultaneously and try to update lambda configuration at the
+    # same time. For simplicity, we don't do any sychronization between them, and simply
+    # retry on conflict, as these updates should be idempotent anyway.
+    num_retries_on_conflict = 3
+
+    for retry in range(num_retries_on_conflict):
+        try:
+            return _maybe_update_lambda(
```

Comment:
I thought we don't need this code segment anymore?

Reply:
We do since there may be a race condition where multiple branches of foreach try to create the same lambda simultaneously. It is much simpler than the previous version though.

Reply:
Is this still needed after the move to `package_init`?